### ISSUE TITLE
xbcomp.exe Virus Alert

### ISSUE THREAD

**Issue by FlyinFistOfJuda** [marchaesen/vcxsrv#26]
Hello,

I updated VcXsrv via winget from the following source to 1.20.14.0:

Name ID Version Source
VcXsrv marha.VcXsrv 1.20.14.0 winget

Defender blocked the xbcomp.exe as Trojan and blacklisted the binary.
Detailed Report from virustotal.com:
[https://www.virustotal.com/gui/file/5e76de6e07913392c7f5d20a0b63305744c784bf1743a2b1c36e5dc9a7ba35e9/detection](https://github.com/ArcticaProject/vcxsrv/issues/virustotal)

Is this a false positive?

Best Regards.

**Comment by EnriqueSLHC** [comment]
I got it also. 

![image](https://github.com/user-attachments/assets/fcc3acbc-afba-4fb1-95aa-5efb124e6654)

however, i was able to get the debug version.

**Comment by JeffsRealm** [comment]
Adding to this, this is now being flagged by multiple antivirus vendors as Malicious. Defender, Avast etc.

**Comment by JeffsRealm** [comment]
So I submitted for Deep Analysis as to why this is getting flagged, somewhere packed inside is a file either created when run or found inside named nshB884.tmp which has no hash signature or the hash for the file is off or unretrievable. 

The file has a signature close to a virus they are suspecting it is because it can not be verified.  And it is doing some sketchy behavior. Querying a whole lot of system environment variables. from the registry. It fine if they need them but some of these I even question, why does this need them?

Some of those are grabbing computer name not sure why the app needs that
then it is grabbing a whole bunch of file explorer settings which are also sketchy, wants to know if files are hidden, if superhidden is enabled if showing internet connection is enabled and so on. So yeah that's what's going on and why it is flagged. 

--

In short there is a packed file inside, when ran becomes some temp file that's trying to execute, the file is unsigned or the signature is corrupt, then this file is querying a whole bunch of computer settings and policy setting about the computer name and how explorer is set to hide various things. It doesn't have this has a Virus perse but highly suspicious activity. Again when I ran it for deep analysis I sat there going why does this app need to know how certain files are hidden and set to hide Or where users might redirect certain personal directories. Let the OS handle it, that's what it is there for. You shouldn't be handling that in your code

I have kind of been watching to make sure we do not have anything going on, but over the weekend I have seen Avast, MCAfee, Defender, and Symantic flag this file as virus, some have let go but others still aren't sure. I believe this is a false positive, but still not letting it run in my org. It's kind of sketchy to me why it needs to query 40 different security policy settings in explorer. I do not see it trying to set them or report them or anything like that, nor any of my analytic show this writing files using any evasion, but could be and are used by malware for evasion techniques. 

My guess is somewhere some code was grabbed or included in this repo that does this and none of this is really needed or they only needed one or two things but the code gets a whole lot more. Then some virus/malware writer somewhere grabbed the same chunk of open sourced code. Being this is unsigned and has same code signature for this behavior as some virus, it is getting caught up. Could even be some malware came and stole a chunk of code from this repo that is getting all these security policy settings and put it in a virus or malware.  That's as much as I can piece together. 

the specific product it is flagging is
Product: VcXsrv
Description: VcXsrv windows xserver
MD5: e1c6b939bda5129726d08760944d2060 
SHA-1: 6fe7399641d18acfb5b6f2a282c156e8293a5d7a 
SHA-256: deca7bf4bc321cdb955efb143ad131913ae30df55cac4318d85bc0d6b11aadfd

**Comment by DakotaNelson** [comment]
FWIW the VirusTotal [sandboxes are failing](https://vtbehaviour.commondatastorage.googleapis.com/5e76de6e07913392c7f5d20a0b63305744c784bf1743a2b1c36e5dc9a7ba35e9_Zenbox.html?GoogleAccessId=758681729565-rc7fgq07icj8c9dm2gi34a4cckv235v1@developer.gserviceaccount.com&Expires=1724089485&Signature=Ktia3MKrIprw98cOgd%2BIwQl8Ym10D%2FJ%2Fe699O9aRIc1sz%2FezzGRJN%2F7pBln6usRQ7aq6ZgfRybKGCZKke38MD2ScqHvKjm3cjYO3ZSDTuq26kHUyvx%2FdZlziWuRRFGD%2FuLzbwtBU81zCWVU1uo0nTyxJQGc1bOpGYy9GMSsz4CGLgJXn2qnl%2BKHk4SMefLpBXbtlVsXM8J1QQVRqc1gRwn%2F4t0b1%2FpxyQl%2BKI8YQUIaDSBEYU8s0qC%2B6yALdzpSUY2SyZnRzQqxCLCLpwlQQ0Vl467w3ugaUPnuGchD2lbZphkKHq7PaKvQLF72QWZ%2FyMKKwFZtbDw8rl3vSp1at6A%3D%3D&response-content-type=text%2Fhtml;#screenshots) because they don't have `libx11.dll` installed, so there really isn't much signal there.

As far as network connections, looks like only microsoft/OCSP and NBNS, both of which make sense - at least in the limited analysis VT can provide.

tl;dr this is probably a false positive but I'm concluding that based on a *lack* of information, because VirusTotal really doesn't have much of anything to go on here.

**Comment by golfromeo-fr** [comment]
I would suggest to add a link to this issue to the Releases section so noone reports this as shady to github. I can't download any version from the github link either the release of February or May or with the scoop tool. Windows Defender rejects as a virus. I really hope you can fix the issue for the next release :-)

**Comment by marchaesen** [comment]
I have no idea how to solve this issue.
xkbcomp is compiled for the official sources (https://gitlab.freedesktop.org/xorg/app/xkbcomp) with some minor modifications.
All the released binaries are compiled from the sources in this repository with Visual Studio community edition. No other binaries are used.

**Comment by efa** [comment]
I think can help to have a toolchain working for MinGW64/MSYS2 [1] that is a very common solution to generate Windows 64bit binaries of opensource projects. So maybe many developers can try and self-build the sources, and so can complaint with antivirus companies.

[1] https://www.msys2.org/
https://en.wikipedia.org/wiki/Mingw-w64

**Comment by FlyinFistOfJuda** [comment]
Currently, only five of the 70Something Antivirus-Software Distributors still detect this binary as a Trojan acc. to Virustotal.
So i guess the main reason for this ticket is "solved" for me, the (forced-upon-me by Corp.-Admins) Defender is no longer quarantaining the binary. I hope everyone is fine with closing this ticket.

**Comment by efa** [comment]
my company PC Defender today still say:
```
Rilevato: Trojan:Win32/Vigorf.A
Stato: Errore
Questa minaccia o app potrebbe non essere completamente risolta.

Data: 15/09/2024 00:25
Dettagli: Questo programma è pericoloso ed esegue comandi ricevuti dall'autore dell'attacco.

Elementi interessati: file: D:\Download\Xserver\VcXsrv\21.1.13\Win64\vcxsrv-64.21.1.13.0.installer.exe
```

**Comment by rask** [comment]
Attempting to run the May release installer .exe brings up a Windows prompt that says it is probable malware, though it would let me run it if I wanted to.

**Comment by onkelbeh** [comment]
Also Trellix blocks all installers from this Release.

**Comment by NeverBeLazyG** [comment]
maybe it has a real virus in it

**Comment by starfishpatkhoo** [comment]
OK, so I have tried to check it in a sandbox, and TL;DR, seems like Windows Defender is triggered by the installer exe. If you either:

1) Install it to C:\Program Files\VcXsrv
2) Unzip the installer.exe with 7z 

And then scan those files, Windows Defender says it is fine. But if you scan the installer itself, then it says there is a Trojan. This is with Windows Defender signatures as of mid November 2024.

So if you want to be safe, unzip the exe with 7z / Sandbox use the CLI.

**Comment by Robbson** [comment]
> So if you want to be safe, unzip the exe with 7z / Sandbox use the CLI.

There is no way to unzip anything if your antivirus tooling blocks this download in the first place.
And in corporate environments you don't have any rights to white list the installer.
So there is still no solution to this issue.

**Comment by starfishpatkhoo** [comment]
Try downloading and unzipping it (or compiling it) on a different computer and transferring it via USB.

If that's not possible, then maybe you need to check with your corporate IT department to make some exceptions, or to do it for you.

If that's not possible, then perhaps you need to ask your corporate IT department for an alternative tool that you can use to do the task / job you're trying to do which requires a Windows-based X-server. There are some paid tools that do the same thing with corporate management features that might be acceptable.

I used to work in Corporate IT, and policy is policy, but if the tool really saves a lot of employee time and can be safely deployed and managed, IT departments are not robots executing policy blindly.. ^_^

**Comment by trev-dev** [comment]
> Try downloading and unzipping it (or compiling it) on a different computer and transferring it via USB.
> 
> If that's not possible, then maybe you need to check with your corporate IT department to make some exceptions, or to do it for you.
> 
> If that's not possible, then perhaps you need to ask your corporate IT department for an alternative tool that you can use to do the task / job you're trying to do which requires a Windows-based X-server. There are some paid tools that do the same thing with corporate management features that might be acceptable.
> 
> I used to work in Corporate IT, and policy is policy, but if the tool really saves a lot of employee time and can be safely deployed and managed, IT departments are not robots executing policy blindly.. ^_^

The way I see it is it's hard enough to keep track of any of the free software I use on a day-to-day basis.  Quite frankly, for me to be personally responsible for how all my applications run on my free system, I would need to abandon most other efforts. 

Antivirus exists for a reason.  We take for granted every day that "somebody else" ensures our free software is under close observation.  This is how stuff like `xz` gets exploited.  I would sooner defer to whomever maintains this project to alleviate the pressure from antivirus than I would try to convince my employer that vcxsrv is worth it.  After all, if you're using Windows 10 or better, you already have WSLg.

**Comment by efa** [comment]
but WSLg work only for X app inside WSL not for other X app coming remotely to the local server. Se we definitively need a general free X server

**Comment by starfishpatkhoo** [comment]
> Antivirus exists for a reason. We take for granted every day that "somebody else" ensures our free software is under close observation. This is how stuff like `xz` gets exploited. I would sooner defer to whomever maintains this project to alleviate the pressure from antivirus than I would try to convince my employer that vcxsrv is worth it. 

Well, as I said, it is only the installer that triggers the Antrivirus, individual files within the installer do not trigger an AV warning. Meaning, the AV is probably catching a false positive. So your best best is still to unzip the installer manually. Try downloading it at home, unzipping it, re-zipping it, upload to Google drive/One drive/Dropbox/email/whatever, download on your office computer, unzip it, and copy and use.

Ultimately, this is open source, you are free to download and compile all the code yourself to prove that the software is not virus infected. In that way, the "somebody else" is you and you do not need to defer to anyone else or worry about any `xv` style supply chain issues.

But if **NONE** of these options are possible for you, then you will have to follow some other people's restrictions - be it from corporate IT, or from some antivirus software's false positives. 

Or use a different product. There is another Windows-based X Server software out there that has also been around for many, many years, so maybe you should try that one instead. ^_^

**Comment by syockit** [comment]
> but WSLg work only for X app inside WSL not for other X app coming remotely to the local server. Se we definitively need a general free X server

If the remote server has ForwardX11 enabled, you can connect with X11 forwarding. See also https://github.com/microsoft/WSL/issues/8367#issuecomment-1138792802

**Comment by efa** [comment]
> If the remote server has ForwardX11 enabled, you can connect with X11 forwarding. See also [microsoft/WSL#8367 (comment)](https://github.com/microsoft/WSL/issues/8367#issuecomment-1138792802)

you are right, in that configuration connecting from local WSL to a remote server, exporting the DISPLAY locally in the SSH tunnel work.

My use case is connecting to a remote server not from WSL but, for example, from PuTTY, setting "Enable X11 forwaring", will set DISPLAY=:10 but without a local Xserver and also with a running local WSL, the X11 applications say:
```
~$ xeyes
PuTTY X11 proxy: unable to connect to forwarded X server: Network error: Connection refused
Error: Can't open display: localhost:10.0
```
This is why I'm still using VcXsrv

![Image](https://github.com/user-attachments/assets/e36e61b9-7438-464d-a4ee-e5179e925af5)

**Comment by starfishpatkhoo** [comment]
> If the remote server has ForwardX11 enabled, you can connect with X11 forwarding. See also [microsoft/WSL#8367 (comment)](https://github.com/microsoft/WSL/issues/8367#issuecomment-1138792802)

I think this is definitely a very useful/good/easy way to go about this.

However, on some machines, I have WSL deployed and used, but on others, I do not need WSL ..

On those other machines, installing an entire Linux Distro into WSL to enable WSLg to run an X server is pretty heavy if you do not use that distro for anything else. VcXserver unzipped is less than 90MB. And portable to even, running off a USB drive. So it is really much, much lighter than using WSLg.

**Comment by StevenLooman** [comment]
Nobody seems to have mentioned that the build process is currently not visible. @marchaesen seems to do this locally on their own computer. We do not see what is happening there. I am not trying to say @marchaesen does something malicious! It might be possible that the environment is infected by another actor, injecting something into the installer during/after the build. Again, I am not trying to say @marchaesen actually does this, or their environment is actually infected.

It would be more transparant if the build is done using, for example, GitHub actions. This way the build is for everybody visible and be repeatable.

**Comment by wswind** [comment]
This project requires Visual Studio 2012 Community Edition to build, and I don't think it uses MinGW. 
The build tools are not natively available on GitHub Actions' hosted runners.

Maybe a zip file release with all files instead of exe installer would be nice for me.

**Comment by starfishpatkhoo** [comment]
Yes, I too prefer a zip file instead of an installer, regardless of AV warning or not. In the mean time, if you are alright with doing some stuff manually/on your own, just use Windows Sandbox to unpack it, AV scan all the files, and then make your own zip of the unpacked files. Then manage the firewall settings yourself, start/stop deamon yourself.. etc.. It does work..

It works fine in a portable manner, so that's how I deploy mine. I've been doing that for a long time now.

I don't mind uploading the zip file somewhere for people to download, but then if you can't trust the download straight from the source repo, why would you ever trust some URL link to a download from some *guy* (?) on the internet comments... LOL... 😆 

Hopefully @marchaesen will keep this in mind when he has some time to come back to this project 😄 In the mean time, thanks a lot for even making the patches for a VC++ version and building the binaries .. I use it daily and I appreciate it.

**Comment by golfromeo-fr** [comment]
> to do this locally on their own computer.

I have tried (and failed) to do the build with docker and VS, I wanted to know if the issue is reproduceable. The build depends on so many software closed sources that are evolving frequently that the automation is going to be difficult, sadly. But I am not an expert so I may have missed a point.

**Comment by wswind** [comment]
The reason I mentioned that it's preferable to have a zip archive is mainly because I've seen some projects, similar to this one, that were packaged using NSIS and subsequently flagged by Windows Defender or other antivirus software. The issue might very well lie with the packaging software itself.

**Comment by Stamgastje** [comment]
I had the same issue before (https://github.com/marchaesen/vcxsrv/issues/29).

But today I downloaded and installed the latest release (21.1.16) and the issue is gone.

**Comment by efa** [comment]
just updated to:
Security Intelligence Update for Microsoft Defender Antivirus -KB2267602 (Version 1.423.270.0) - Current Channel (Generic)
and:
vcxsrv-64.21.1.16.0.installer.exe
give same warn, our IT block the execution of the binary.

Work well unzipping the files and overwrite the ones in:
C:\Program Files\VcXsrv

As sayd, will help to have a simple zip of files without the embedded installer

**Comment by Saxomania** [comment]
i am still on Version 1.423.266.0 and Defender did not block it - has been blocked before everytime i tried it (but tried it just every now and then)

**Comment by FlyinFistOfJuda** [comment]
I'm closing this issue. It's old and we've got newer versions.
